biz.aQute.bndlib/src/aQute/bnd/plugin/spi/SPIDescriptorGenerator.java

```diff
@@ -56,18 +58,40 @@ public void verify(Analyzer analyzer) throws Exception {
 		for (Entry<String, ArrayList<String>> entry : providerTypes.entrySet()) {
 			String key = "META-INF/services/" + entry.getKey();
 			ArrayList<String> list = entry.getValue();
-			list.add(0, "# Generated by bnd (modified: " + new Date(analyzer.lastModified()) + ")");
-			String value = Strings.join("\n", entry.getValue());
 
 			Resource resource = analyzer.getJar()
 				.getResource(key);
 
+			String value = "";
+
 			if (resource != null) {
-				value = IO.collect(resource.openInputStream()) + "\n" + value;
+				value = IO.collect(resource.openInputStream());
+
+				try (BufferedReader br = new BufferedReader(new StringReader(value))) {
+					String line;
+					while ((line = br.readLine()) != null) {
+						line = line.trim();
+						if (line.isEmpty() || (line.charAt(0) == '#')) {
+							continue;
+						}
+						list.remove(line);
+					}
+				}
 			}
 
-			analyzer.getJar()
-				.putResource(key, new EmbeddedResource(value, analyzer.lastModified()));
+			// Only write if we have something to add
+			if (!list.isEmpty()) {
```

Comment:
Shouldn't the BufferedReader loop also be in this if? If there is nothing to add, why do anything here?

Reply:
The Map.Entry won't exist if there's nothing in this list (since an entry is only added when there is a list to add).

Later on, however, we may have removed all entries from the list if they already existed in the descriptor.

Reply:
does this explanation make sense?